### PR TITLE
mpi4py 3.0.2 (new formula)

### PR DESCRIPTION
**Pull request by ax3l**
Adds `mpi4py` which are popular Cython bindings for MPI used in many Python projects (h5py, adios, adios2, ...).

Dependency for #45997

CC-ing upstream maintainer @dalcinl .

- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?
(`brew audit` install is broken since last update today)
-----

### CODE REVIEW

File: Formula/mpi4py.rb
Comment (ax3l) [inline Formula/mpi4py.rb]: note: instructions taken from `numpy.rb`

**Comment by zbeekman** [comment]
It looks like there's a bug in `brew audit --online --new-formula` effecting Catalina and High Sierra:

```
       Error: undefined method `[]' for nil:NilClass
/usr/local/Homebrew/Library/Homebrew/dev-cmd/audit.rb:621:in `audit_bitbucket_repository'
/usr/local/Homebrew/Library/Homebrew/dev-cmd/audit.rb:1006:in `block in audit'
/usr/local/Homebrew/Library/Homebrew/dev-cmd/audit.rb:999:in `each'
/usr/local/Homebrew/Library/Homebrew/dev-cmd/audit.rb:999:in `audit'
/usr/local/Homebrew/Library/Homebrew/dev-cmd/audit.rb:123:in `block in audit'
/usr/local/Homebrew/Library/Homebrew/dev-cmd/audit.rb:116:in `each'
/usr/local/Homebrew/Library/Homebrew/dev-cmd/audit.rb:116:in `audit'
/usr/local/Homebrew/Library/Homebrew/brew.rb:102:in `<main>'
```

mpi4py has no audit errors on mojave

**Review by zbeekman — APPROVED** [review]
LGTM

**Comment by zbeekman** [comment]
I'll let this sit a little bit to see if other maintainers have strong opinions.

**Comment by zbeekman** [comment]
As far as I can tell there is a bug with audit for bitbucket repositories on catalina and high-sierra.

File: Formula/mpi4py.rb
Comment (zbeekman) [inline Formula/mpi4py.rb]: Please get rid of the head block, unless you have a good reason for keeping it

File: Formula/mpi4py.rb
Comment (zbeekman) [inline Formula/mpi4py.rb]: Can you please improve the tests here a little bit? Just importing the module may not be sufficient. Using it (briefly) would be nice. Look at numpy's test for ideas.

**Review by zbeekman — CHANGES_REQUESTED** [review]

**Comment by zbeekman** [comment]
FYI, I opened a bug report for the audit issue in Homebrew/brew#6681

File: Formula/mpi4py.rb
Comment (dalcinl) [reply Formula/mpi4py.rb]: If importing is not enough, I suggest the following quick testing, both of them (even the `helloworld` one) exercise basic MPI point-to-point communication. Testing beyond that may prove unreliable, trust me, do not open a can or worms...

```
$ mpiexec -n 4 python3 -m mpi4py.run -m mpi4py.bench helloworld
Hello, World! I am process 0 of 4 on localhost.
Hello, World! I am process 1 of 4 on localhost.
Hello, World! I am process 2 of 4 on localhost.
Hello, World! I am process 3 of 4 on localhost.

$ mpiexec -n 4 python3 -m mpi4py.run -m mpi4py.bench ringtest -l 10 -n 1024
time for 10 loops = 5.55515e-05 seconds (4 processes, 1024 bytes)
```

Comment (zbeekman) [reply Formula/mpi4py.rb]: I agree, we don't need extensive testing, just confirmation that mpi4py is able to use Homebrew's open-mpi

File: Formula/mpi4py.rb
Comment (ax3l) [reply Formula/mpi4py.rb]: I was using head of mpi4py for several months until the 3.0.0 release was tagged, sometimes upstream MPI flavors have breaking changes that need adoption from nightly.
But can remove it, currently all is good.

File: Formula/mpi4py.rb
Comment (ax3l) [reply Formula/mpi4py.rb]: btw, how flexible is this package regarding OpenMPI/MPICH and is there a reason for the current preference?

File: Formula/mpi4py.rb
Comment (zbeekman) [reply Formula/mpi4py.rb]: Ok, it sounds like maybe this would be ok to keep? 

If upstream (@dalcinl) added a nightly `brew install --HEAD mpi4py && brew test --HEAD mpi4py` test I’d definitely give it a thumbs up. 

But if you anticipate using/needing it then that could be ok.

File: Formula/mpi4py.rb
Comment (zbeekman) [reply Formula/mpi4py.rb]: Not 100% sure what you mean. Homebrew defaults to Open-MPI only so that all MPI packages will play nice together. I started to put together MPICH tooling over at the unofficial brew-sci tap but haven’t made much progress.

Comment (zbeekman) [reply Formula/mpi4py.rb]: Not 100% sure what you mean. Homebrew defaults to Open-MPI only so that all MPI packages will play nice together. I started to put together MPICH tooling over at the unofficial brew-sci tap but haven’t made much progress.

**Review by zbeekman — APPROVED** [review]
LGTM

**Comment by zbeekman** [comment]
I'll merge this so as not to hold up adios any longer. If you want to add HEAD capability, please open a new PR and we can look/discuss there. Thanks!

**Comment by zbeekman** [comment]
Thanks for your contribution to Homebrew! 🎉 🥇 

Without awesome contributors like you, it would be impossible to maintain Homebrew to the high level of quality users have come to expect. Thank you @ax3l!

File: Formula/mpi4py.rb
Comment (ax3l) [reply Formula/mpi4py.rb]: Okay, let's add this when needed I guess :)

**Comment by ax3l** [comment]
Thank you for you great guidance and review, @zbeekman! Definitely enjoyed contributing.

**Comment by ax3l** [comment]
Hi @zbeekman, quick beginners question: I noticed the bot has already added macOS bottles but no Linux x86_64_linux bottle yet. Do I have to trigger something for the latter or is this just in the backlog?

**Comment by issyl0** [comment]
@ax3l Linux support is done in a separate repo, [Homebrew/linuxbrew-core](https://github.com/Homebrew/linuxbrew-core). As per that [contributing guide](https://github.com/Homebrew/linuxbrew-core/blob/master/CONTRIBUTING.md#contributing-to-linuxbrew), changes are merged in usually daily, then we build bottles for changed formulae. If you don't see a bottle PR within the next couple of days (though of course, now I'm looking out for it!), feel free to [raise a "bottle request" issue](https://github.com/Homebrew/linuxbrew-core/issues/new/choose) and give us a nudge. Thanks!

**Pull request closed without merging**
